akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolGateway.scala

```diff
@@ -4,91 +4,85 @@
- * provided by its [[PoolInterfaceActor]] actor. The actor accepts [[PoolInterfaceActor.PoolRequest]] messages
- * and completes their `responsePromise` whenever the respective response has been received (or an error occurred).
+ * Manages access to a host connection pool through the [[PoolMasterActor]]
  *
- * A [[PoolGateway]] provides a layer of indirection between the pool cache and the actual
- * pools that is required to allow a pool incarnation to fully terminate (e.g. after an idle-timeout)
- * and be transparently replaced by a new incarnation if required.
- * Removal of cache entries for terminated pools is also supported, because old gateway references that
- * get reused will automatically forward requests directed at them to the latest pool incarnation from the cache.
+ * A [[PoolGateway]] is represented by its [[HostConnectionPoolSetup]] and its [[GatewayIdentifier]]. If the later
+ * is [[SharedGateway]], it means that a shared pool must be used for this particular [[HostConnectionPoolSetup]].
  */
-private[http] class PoolGateway(hcps: HostConnectionPoolSetup,
-                                _shutdownStartedPromise: Promise[Done])( // constructor arg only
-                                  implicit system: ActorSystem, fm: Materializer) {
-  import PoolGateway._
-  import fm.executionContext
+private[http] final class PoolGateway(gatewayRef: ActorRef, val hcps: HostConnectionPoolSetup, val gatewayId: GatewayIdentifier)(implicit fm: Materializer) {
+
+  /**
+   * Send a request through the corresponding pool.
+   *
+   * @param request the request
+   * @return the response
+   */
+  def apply(request: HttpRequest): Future[HttpResponse] = {
+    val responsePromise = Promise[HttpResponse]()
+    gatewayRef ! SendRequest(this, request, responsePromise, fm)
+    responsePromise.future
+  }
 
-  private val state = {
+  /**
+   * Start the corresponding pool to make it ready to serve requests.
+   *
+   * @return the gateway itself
+   */
+  def startPool(): PoolGateway = {
+    gatewayRef ! StartPool(this, fm)
+    this
+  }
+
+  /**
+   * Shutdown the corresponding pool and signal its termination. The pool can be restarted by sending a new request
+   * through this gateway.
+   *
+   * @return a Future completed when the pool has been shutdown.
+   */
+  def shutdown(): Future[Done] = {
     val shutdownCompletedPromise = Promise[Done]()
-    val props = Props(new PoolInterfaceActor(hcps, shutdownCompletedPromise, this)).withDeploy(Deploy.local)
-    val ref = system.actorOf(props, PoolInterfaceActor.name.next())
-    new AtomicReference[State](Running(ref, _shutdownStartedPromise, shutdownCompletedPromise))
+    gatewayRef ! Shutdown(this, shutdownCompletedPromise)
+    shutdownCompletedPromise.future
+  }
+
+  override def toString = s"PoolGateway(hcps = $hcps)"
+
+  // Testing only.
+  def poolStatus(): Future[Option[PoolInterfaceStatus]] = {
```

Comment:
Yes. I can restrict it to `private[client]`, and `Http().poolSize()` to `private[scaladsl]`.